phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,550 @@
+                + cmdLine.getOptionValue(PORT_OPTION.getOpt());
+        logFile = cmdLine.getOptionValue(LOG_FILE_OPTION.getOpt());
+        inputFile = cmdLine.getOptionValue(TABLE_CSV_FILE_OPTION.getOpt());
+        dryRun = cmdLine.hasOption(DRY_RUN_OPTION.getOpt());
+        compaction = cmdLine.hasOption(DRY_RUN_OPTION.getOpt());
+    }
+
+    @VisibleForTesting
+    public void prepareToolSetup() {
+        try {
+            if (logFile != null) {
+                FileHandler fh = new FileHandler(logFile);
+                fh.setFormatter(new SimpleFormatter());
+                LOGGER.addHandler(fh);
+            }
+
+            prop.put(Indexer.INDEX_BUILDER_CONF_KEY, PhoenixIndexBuilder.class.getName());
+            prop.put(NonTxIndexBuilder.CODEC_CLASS_NAME_KEY, PhoenixIndexCodec.class.getName());
+
+            if (inputTables == null) {
+                inputTables = new String(Files.readAllBytes(Paths.get(inputFile)));
+            }
+            if (inputTables == null) {
+                LOGGER.severe("tables' list is not available; use -tb or -f option");
+                System.exit(1);
+            }
+            LOGGER.info("list of tables/indexes passed: " + inputTables);
+
+            tablesAndIndexes = extractTablesAndIndexes(inputTables);
+
+            if (operation.equalsIgnoreCase(UPGRADE_OP)) {
+                upgrade = true;
+            } else if (operation.equalsIgnoreCase(ROLLBACK_OP)) {
+                upgrade = false;
+            } else {
+                throw new IllegalStateException("Invalid option provided for "
+                        + OPERATION_OPTION.getOpt() + " expected values: {upgrade, rollback}");
+            }
+            if (dryRun) {
+                LOGGER.info("This is the beginning of the tool with dry run.");
+            }
+        } catch (IOException e) {
+            LOGGER.severe("Something went wrong "+e);
+        }
+    }
+
+    @VisibleForTesting
+    public void executeTool() {
+        try (Connection conn = DriverManager.getConnection(connectionURL, new Properties())) {
+            ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class)
+                    .getQueryServices();
+            Configuration conf = queryServices.getConfiguration();
+            executeTool(conn, queryServices, conf);
+        } catch (SQLException e) {
+            LOGGER.severe("Something went wrong in executing tool "+ e);
+        }
+    }
+
+    private void executeTool(Connection conn, ConnectionQueryServices queryServices,
+            Configuration conf) {
+
+        LOGGER.info("Executing "+ operation);
+
+        for (Map.Entry<String, ArrayList<String>> entry: tablesAndIndexes.entrySet()) {
+            String dataTableFullName = entry.getKey();
+            ArrayList<String> indexes = entry.getValue();
+
+            try (Admin admin = queryServices.getAdmin()) {
+
+                PTable dataTable = PhoenixRuntime.getTableNoCache(conn, dataTableFullName);
```

Comment:
This logic in general is assuming that the logical table name in Phoenix == the physical table in HBase. This assumption's not always going to be valid. 

Three big cases: 
1. If namespaces are enabled in HBase, and namespace / schema mapping is enabled in Phoenix. In that case, a table named FOO.BAR in Phoenix will be named FOO:BAR in HBase
2. Global views can own indexes, and global views will have different names from the physical table they're located in. Likewise, indexes on views will have names in Phoenix, but be physically located in tables named _IDX_<name of HBase data table> 
3. Tenant-owned views can own indexes, and they can't be seen using a PhoenixRuntime.getTable() command, unfortunately.

Reply:
Thinking about this more -- I think for 2 and 3 we just need to automatically upgrade the _IDX_ table when its associated data table is being upgraded, and downgrade it with the associated table too. Because all view indexes of a table share the same physical table, we can't pick and choose like we do for global indexes of a base table.

Reply:
In Point 2 and 3, the tool accepts the indexes and table list to be upgraded. For consistency, it is better to pass all the indexes as you said. In that case, I think only point one needs to be addressed here. Is that correct?

Reply:
@swaroopak , yes, I think just automatically doing all view indexes of a data table passed in would be best. But you'll have to add logic to map the logical view index names to the physical view index table (and make sure you only do the upgrade once) if they provide view index names. 

There should also be error handling logic if the user provides data table and/or index names that don't exist. 

Another safety feature -- we should make sure that any index name passed in is a global index and not a local one. Local indexes are unaffected by the new global index design.

Reply:
done.

Reply:
I am thinking about how to check if the view index is local or global. ll update the PR as per my finding. meanwhile, I fixed the rest of the comments on this thread.

Reply:
Since the new logic will only return global indexes on a view, I don't need to have this additional check -- resolving the comment.